warmrestart/warm_restart.cpp

```diff
@@ -1,108 +1,139 @@
 {
-    {INIT,          "init"},
+    {INITIALIZED,   "initialized"},
     {RESTORED,      "restored"},
     {RECONCILED,    "reconciled"}
 };
 
-WarmStart &WarmStart::getInstance()
+WarmStart &WarmStart::getInstance(void)
 {
     static WarmStart m_warmStart;
     return m_warmStart;
 }
 
+/*
+ * WarmStart's initialization method -- to be invoked once per application.
+ */
+void WarmStart::initialize(const std::string &app_name,
+                           const std::string &docker_name)
+{
+    auto& warmStart = getInstance();
+
+    if (warmStart.m_initialized)
+    {
+        return;
+    }
+
+    warmStart.m_stateDb =
+        std::make_shared<swss::DBConnector>(STATE_DB, swss::DBConnector::DEFAULT_UNIXSOCKET, 0);
+
+    warmStart.m_stateWarmRestartTable =
+        std::unique_ptr<Table>(new Table(warmStart.m_stateDb.get(), STATE_WARM_RESTART_TABLE_NAME));
+
+    warmStart.m_cfgDb =
+        std::make_shared<swss::DBConnector>(CONFIG_DB, swss::DBConnector::DEFAULT_UNIXSOCKET, 0);
+
+    warmStart.m_cfgWarmRestartTable =
+        std::unique_ptr<Table>(new Table(warmStart.m_cfgDb.get(), CFG_WARM_RESTART_TABLE_NAME));
+
+    setWarmStartState(app_name, WarmStart::INITIALIZED);
+
+    warmStart.m_initialized = true;
+}
+
 /*
  * <1> Upon system reboot, the system enable knob will be checked.
  * If enabled, database data will be preserved, if not, database will be flushed.
  * No need to check docker level knobs in this case since the whole system is being rebooted .
 
  * <2> Upon docker service start, first to check system knob.
- * if enabled, docker warm start should be performed, otherwise system warm reboot will be ruined.
- * If system knob disabled, while docker knob enabled, this is likely an individual docker warm restart request.
+ * if enabled, docker warm-start should be performed, otherwise system warm-reboot will be ruined.
+ * If system knob disabled, while docker knob enabled, this is likely an individual docker
+ * warm-restart request.
 
  * Within each application which should take care warm start case,
  * when the system knob or docker knob enabled, we do further check on the
- * actual warm start state ( restart_count), if no warm start state data available,
- * the database has been flushed, do cold start. Otherwise warm start.
+ * actual warm-start state ( restore_count), if no warm-start state data available,
+ * the database has been flushed, do cold start. Otherwise warm-start.
  */
 
-// Check warm start flag at the very begining of application, do it once for each process.
-bool WarmStart::checkWarmStart(const std::string &app_name, const std::string &docker_name)
+/*
+ * Method to verify/obtain the state of Warm-Restart feature for any warm-reboot
+ * capable component. Typically, this function will be called during initialization of
+ * SONiC modules; however, method could be invoked at any given point to verify the
+ * latest state of Warm-Restart functionality.
```

Comment:
The comment here is a little missleading.
 ```
however, method could be invoked at any given point to verify the
 * latest state of Warm-Restart functionality.
```
The function also updates the restart_count whenever called, not just perform check.

Reply:
Ok, will adjust that comment.

Reply:
done